### PR TITLE
Refactor/hexview

### PR DESCRIPTION
**Pull request by dannyp303**
- [X] I have reviewed the [OFRAK contributor guide](https://ofrak.com/docs/contributor-guide/getting-started.html) and attest that this pull request is in accordance with it.

**One sentence summary of this PR (This should go in the CHANGELOG!)**
- Refactor HexView and related components to use mousewheel instead of scroll and compartmentalize all comonents to src/hex. 
**Link to Related Issue(s)**

**Please describe the changes in your request.**

**Anyone you think should look at this, specifically?**

### CODE REVIEW

File: frontend/src/hex/ByteclassView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/ByteclassView.svelte:134]: Why does this no longer need to be divided by two?

File: frontend/src/hex/HexView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/HexView.svelte:90]:
```suggestion
```

Does this need to be here?

File: frontend/src/hex/HexView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/HexView.svelte:256]:
```suggestion
    if (dataSearchResults.matches != null) {
```

File: frontend/src/hex/HexView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/HexView.svelte:285]: The more "correct" way to do this in Svelte is to `bind:this` on the `scrollable` element. The variable should be reactively updated on mount.

File: frontend/src/hex/MinimapView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/MinimapView.svelte:5]:
```suggestion
    height: 75vh;
```

Why should it only be 75% of the view height?

File: frontend/src/utils/Pane.svelte
Comment (rbs-jacob) [inline frontend/src/utils/Pane.svelte:17]: This may have fixed some short-term problems, but I am really skeptical that this is not the cause of a bunch of other scroll things breaking.

I don't think this should have overflow hidden.

**Review by rbs-jacob — CHANGES_REQUESTED** [review]
Some outstanding issue that should be fixed before this can be merged.

- [x] Likely due to `Pane` refactor, this PR removes padding all over the GUI that should be fixed
- [x] The search bar in the tree view is no longer pinned to the top, and scrolls away when you scroll the tree
- [x] The minimap now floats half way between the right of the hex view and the right of the screen when it should be pinned to the right of the screen
- [x] The minimap no longer takes up the full height of the screen, which makes it less practical to use, less easy to see details, etc.
- [x] The minimap and "go to offset" input no longer have a black background, and the hex can be seen around/behind them if the hex area is thin enough to cause overlap
- [ ] The "breadcrumb" that used to be at the top of the hex view is no longer visible (it does seem to come back in the disassembly view for some reason)
- [ ] The hex view does not allow scrolling past the bottom when using the scroll wheel – I know there is a comment in the code that this is a bug, but I actually ended up kind of preferring it this way
- [ ] Scrolling with the mini map does not round to the nearest line correctly (see video at the bottom)
- [ ] The bottom of the red "view window" in the minimap is not opaque
- [ ] Scrolling in the disassembly view no longer works
- [ ] Scrolling in the text view no longer works

Before: 
image

After:
image

Scrolling not working right:

https://github.com/redballoonsecurity/ofrak/assets/99368685/94f57fb6-a866-40fe-8d5d-e7b89db6fc5c

Left button bar lack of background:

image

Minimap needs background, and should be full height:

image

File: frontend/src/utils/SearchBar.svelte
Comment (dannyp303) [reply frontend/src/utils/SearchBar.svelte:121]: Search bar takes a callback that tells it how to perform the search, and an object to populate with the results. I needed to clear the search bar from another component, so I did so by clearing the search results.

File: frontend/src/hex/ByteclassView.svelte
Comment (dannyp303) [reply frontend/src/hex/ByteclassView.svelte]: Why did it need to in the first place? 
The screen height should be the number of bytes being displayed as it is calculated with:
`  $screenHeight =
      Math.floor(hexDisplay.offsetHeight / lineHeight) * alignment;`
being the number of lines currently being displayed * 16 bytes per line. i.e bytes per line. 

The height of the rectangle is then the percentage of the current data being displayed * the size of the canvas.

**Review by rbs-jacob — CHANGES_REQUESTED** [review]
Edge case that may not be worth worrying about, but the minimap gets cut off if the hex view gets too small.

https://github.com/redballoonsecurity/ofrak/assets/99368685/97052f59-8010-4887-b698-f36fbb3b3aa4

File: frontend/src/views/ComponentsView.svelte
Comment (rbs-jacob) [inline frontend/src/views/ComponentsView.svelte:108]:
```suggestion
    margin: 1em 0;
```

Let's split the difference and bump it up a little.

File: frontend/src/utils/Pane.svelte
Comment (rbs-jacob) [inline frontend/src/utils/Pane.svelte:23]: Reflecting on our discussion, I think we should have the default be at least `1em`, maybe `0.5em` at the very very least. Then, we can override this value to be `0` for the tree view since that's where it will look the best to have the pane move vertically.

My thinking is that this will make the default for non-tree views in the top left area of the page look much better. For example, there should (in my opinion) be more padding below the buttons at the bottom here:

image

File: frontend/src/ofrak/remote_resource.js
Comment (rbs-jacob) [inline frontend/src/ofrak/remote_resource.js:175]: Is this code necessary?

File: frontend/src/hex/JumpToOffset.svelte
Comment (rbs-jacob) [inline frontend/src/hex/JumpToOffset.svelte:43]: This needs to be rounded. We can now go to an unaligned position like `7` and then the rest of the results will be misaligned when scrolling the hex view.

Also the value no longer clears properly after it has been set. So once it has been set the value no longer updates on scroll.

Last thing is that we should be using the actual value, not the placeholder value. Not only does having full-opacity text look better, but it means that a user can edit the current location instead of having to manually retype it from scratch every time. It would also solve the above problem.


https://github.com/redballoonsecurity/ofrak/assets/99368685/961d85d4-37cc-4d9b-b826-944bdbfcda24

File: frontend/src/hex/HexView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/HexView.svelte:184]:
```suggestion
    if (children.length === 0) {
```

File: frontend/src/hex/HexView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/HexView.svelte:282]: I think we talked about having this be set by `bind:this`?

**Review by rbs-jacob — CHANGES_REQUESTED** [review]

File: frontend/src/hex/HexView.svelte
Comment (dannyp303) [reply frontend/src/hex/HexView.svelte:282]: Woops, i added this, but missed this re-assignment.

File: frontend/src/hex/HexView.svelte
Comment (rbs-jacob) [inline frontend/src/hex/HexView.svelte:125]: This should also be rounded, otherwise search can jump to an unaligned offset.

**Review by rbs-jacob — CHANGES_REQUESTED** [review]
One additional observation about the pane changes is that the vertical and horizontal scrolls are handled differently in a way that I think is potentially confusing for a user. Specifically: there are horizontal scroll bars, but they're way at the bottom and can go out of view. Also if you middle click in the tree with the mouse wheel, it only lets you scroll horizontally. 

Similarly, horizontal scrolling the hex view doesn't work if it's too small. 

I've included a video, but I don't consider fixing the horizontal scroll critical for this PR. If it's an easy fix, we should do it. But it shouldn't block the PR getting merged.

https://github.com/redballoonsecurity/ofrak/assets/99368685/0e673b5f-d744-43a6-9541-701dad530cff

**Comment by dannyp303** [comment]
> One additional observation about the pane changes is that the vertical and horizontal scrolls are handled differently in a way that I think is potentially confusing for a user. Specifically: there are horizontal scroll bars, but they're way at the bottom and can go out of view. Also if you middle click in the tree with the mouse wheel, it only lets you scroll horizontally.
> 
> Similarly, horizontal scrolling the hex view doesn't work if it's too small.
> 
> I've included a video, but I don't consider fixing the horizontal scroll critical for this PR. If it's an easy fix, we should do it. But it shouldn't block the PR getting merged.
> 
>  Screen.Recording.2024-02-27.at.5.37.07.PM.mov

Horizontal scrolling here would be a bit weird, I think the best solution would be to adjust the alignment size based on the hexview element width. In the interest of time, i think this should be a separate PR.

**Comment by rbs-jacob** [comment]
> Horizontal scrolling here would be a bit weird, I think the best solution would be to adjust the alignment size based on the hexview element width. In the interest of time, i think this should be a separate PR.

I'm more concerned about horizontal scrolling in the tree view being weird than about not being able to horizontally scroll the hex view. The only reason I bring them up together is that I think they're likely related. 

Like I said, no need to block this PR on that.

File: frontend/src/hex/JumpToOffset.svelte
Comment (rbs-jacob) [reply frontend/src/hex/JumpToOffset.svelte]: The last part of this one is still unresolved – the go to offset text input should set the `value` and not the `placeholder` so that it looks better and stays updated.

**Comment by dannyp303** [comment]
> > Horizontal scrolling here would be a bit weird, I think the best solution would be to adjust the alignment size based on the hexview element width. In the interest of time, i think this should be a separate PR.
> 
> I'm more concerned about horizontal scrolling in the tree view being weird than about not being able to horizontally scroll the hex view. The only reason I bring them up together is that I think they're likely related.
> 
> Like I said, no need to block this PR on that.


Not related, the hex view doesnt "scroll" so i have scrolling totally turned off for that view, if I enable it, it does the "double scroll" that initiated this entire PR. Regardin the resource view, easy fix, theres an `overflow-x: auto` in the treebox div that isnt required any more.

**Review by rbs-jacob — APPROVED** [review]
Thanks for all of the revisions. Looks good to me! 🤑

**Pull request merged**
